wicker/l5ml_datastore/v2/index_ops.py

```diff
@@ -0,0 +1,253 @@
+            new_fields[name] = field
+    return DatasetSchema(fields=list(new_fields.values()), primary_keys=left.primary_keys)
+
+
+def join_datasets(
+    left: AbstractDataset[Example], right: AbstractDataset[Example], how: str = "outer"
+) -> AbstractDataset[Example]:
+    """Join two datasets (that have the same primary keys) on their primary keys.
+    :param left: First dataset to merge.
+    :param right: Second dataset to merge.
+    :param how: Type of merge to apply. The possible values follow the options for pandas.Dataframe.merge:
+      "left": use only keys from left frame, similar to a SQL left outer join
+      "right": use only keys from right frame, similar to a SQL right outer join
+      "outer": use union of keys from both frames, similar to a SQL full outer join
+      "inner": use intersection of keys from both frames, similar to a SQL inner join
+    :return: A new dataset resulting from the merge of "left" and "right".
+    """
+    # Note, this function does a very simple merge based on equality of keys.
+    # Maybe we should have a function for resolving approximate timestamps.
+    # We could truncate the precision of timestamps, but that's not a very good solution.
+    # Better to use the left side's timestamp as reference and merge right side rows that are within an
+    # error threshold.
+    if not left.is_compatible(right):
+        raise WickerDatastoreException(
+            "Dataset a and b are not compatible. Datasets must be of the same type "
+            "and have the same backing infrastructure."
+        )
+    left_required = False
+    right_required = False
+    if how == "inner":
+        left_required = True
+        right_required = True
+    elif how == "left":
+        left_required = True
+    elif how == "right":
+        right_required = True
+    elif how == "outer":
+        pass
+    else:
+        raise WickerDatastoreException(f"Unsupported join type {how}.")
+    dst_schema = merge_schemas(left.schema, right.schema, left_required, right_required)
+
+    dst_df = left.arrow_table.to_pandas().merge(right.arrow_table.to_pandas(), how=how, on=left.schema.primary_keys)
+    dst_df.sort_index(inplace=True)
+
+    return left.create_compatible_anonymous_dataset(pyarrow.Table.from_pandas(dst_df), dst_schema)
+
+
+class ColumnWindowBuilder:
+    """Collection of helper functions to build window specifications for apply_temporal_window_specifiction."""
+
+    @staticmethod
+    def uniform_history(
+        column_name: str, num_samples: int, delta: int, margin: int, is_required: bool = True
+    ) -> ColumnHistorySpecification:
+        """uniform_history creates a request for regularly spaced objects before the current row.
+        For example:
+        uniform_history("image", 4, 100*MS, 10*MS, is_required=True)
+        requests that for each example we fetch 4 objects from the "image" column that are at
+        0ms, -100ms, -200ms, -300ms relative to the example's row.
```

Comment:
So this is more or less returning the time offsets from one position in the dataset?